Some authentication changes for improving documentation

xi opened the pull request:
This is about improving the documentation on authentication backends and object permissions. See also
#10600 and #10601.

This is work in progress. It is supposed to illustrate current issues and how they could be fixed.

carltongibson commented:
Hi @xi. Thanks for this. It makes it easier to review the combined changes together if we can. 

How good are you with managing your git history? Ideally we want exactly 3 commits here:

* `Fixed #29918 -- Added support for object permissions in PermissionRequiredMixin`.
* `Fixed #20218 -- Added DefaultObjectPermissionsBackend`.
* `Refs #29918, #20218 -- Improved Authentication Docs.`

Are you able to adjust for this?

xi commented:
Hi @carltongibson. I rebased my changes. The history is now slightly different than what you suggested: 
f536a2f  is about general doc improvements that are not related to the further changes. 
fe9a441 is about changes as discussed in [this mailing list thread](https://groups.google.com/forum/#!topic/django-developers/CNmz22gEsCc)

carltongibson commented:
OK, no problem — it's the general idea that matters just now. Thanks!

timgraham commented:
As for me, I'd rather have separate PRs to review and merge separately unless the commits are tightly coupled which doesn't seem to be the case here.

xi commented:
> As for me, I'd rather have separate PRs to review and merge separately unless the commits are tightly coupled which doesn't seem to be the case here.

I think this is something you as reviewers have to decide. For me it is fine either way. Feel free to re-open and comment/merge #10600 and #10601. I can rebase this pull request afterwards.

carltongibson commented:
Sorry @timgraham — I asked to have it in one place. I thought #10600 and #10601 required docs changes; @xi made reasonable points that a bit of a re-write was needed; at that point it became coupled such that I wanted to address it as one. It's in my queue for this week.

carltongibson commented on docs/ref/contrib/auth.txt:
```diff
@@ -629,6 +629,16 @@ The following backends are available in :mod:`django.contrib.auth.backends`:
    users because :attr:`~RemoteUserBackend.user_can_authenticate` always
    returns ``True``.
 
+.. class:: DefaultObjectPermissionsBackend
```
Again, I'd put these changed next to `ModelBackend`.  (`RemoteUserBackend` is pretty niche...) 

* I think in this block you need to include the `AUTHENTICATION_BACKENDS` settings example to show how to actually configure use this: it's important that it's used **in addition** to `ModelBackend`
* The `superset` phrasing assumes too much knowledge: We need an example of the behaviour with the default settings and adding the new backend. Something **along the lines of** "with `DefaultObjectPermissionsBackend` the permission check will fallback to the general permission if the user doesn't have the specific object permission".

carltongibson commented on docs/releases/2.2.txt:
```diff
@@ -57,6 +57,9 @@ Minor features
   checking object permissions.
+* The new
+  :class:`~django.contrib.auth.backends.DefaultObjectPermissionsBackend` makes
+  object permissions a superset of global permissions.
```
Again, the "superset" phrasing is too opaque.

carltongibson commented on docs/topics/auth/customizing.txt:
```diff
@@ -255,15 +269,56 @@ authenticated users do not.
-the check performed). An authentication backend will receive the keyword
-parameters ``obj`` and ``user_obj`` for each object related authorization
-method and can return the object level permission as appropriate.
+there is no implementation for it in the core. An authentication backend will
```
With the combined changes here this line is no longer true. There **is** an implementation in core.

xi replied:
I think that would be an overstatement. `DefaultObjectPermissionObject` is not useful on its own. If you use only `ModelBackend` and `DefaultObjectPermissionObject` you could just as well only use the former and not pass objects.

`DefaultObjectPermissionObject` is interesting if you use one backend handeling global permissions and a second one handeling object permissions.

carltongibson replied:
OK, leave this one.

carltongibson commented on docs/topics/auth/customizing.txt:
```diff
@@ -274,6 +274,15 @@ return ``False`` or an empty set (depending on the check performed).
+    can add the
+    :class:`~django.contrib.auth.backends.DefaultObjectPermissionsBackend` to
+    :setting:`AUTHENTICATION_BACKENDS`. It will return all global permissions
+    when checking for object permissions, resulting in the expected behavior.
```
So at this stage, I think this section just doesn't read. There's two topics: 

1. Handling object permissions with he shipped components. 
2. Creating a custom backend that handles object permissions.

carltongibson replied:
For this kind of thing I imagine the "Advance Beginner" — they obviously got past the basics since they're here. SO someone wanting to implement object permissions and needing to see an explanation of what's provided and what they then have to do.

This was my point about building the examples here (from the default to adding a custom backend).

> I think it is crucial to explain how this can actually be used.

Absolutely! That's what's kind of missing: "how do I use this new stuff?"

(If you want to mention `get_permission_object()`, perhaps after the custom backend, then that makes sense there.)

xi commented on docs/topics/auth/index.txt:
```diff
@@ -44,6 +44,7 @@ of these common problems have been implemented in third-party packages:
 * Password strength checking
 * Throttling of login attempts
 * Authentication against third-parties (OAuth, for example)
+* Fine-grained permissions
```
As far as I understand, the second argument to `has_perm()` is indented to be a model instance, but can theoretically be anything. So I believe the existing system can be used for other things than object-level permissions.

I understand that this is rather vague though. I will change it.

xi commented on docs/topics/auth/customizing.txt:
```diff
@@ -259,12 +259,48 @@ Handling object permissions
+        def get_user(self, user_id):
+            return None
+
+        def get_all_permissions(self, user_obj, obj=None):
```
I am not so sure about the "others can only view". We have no clue here which other authentication backends are active.

Maybe this is nitpicking. I will change it for now.

xi commented on docs/topics/auth/customizing.txt:
```diff
@@ -259,12 +259,48 @@ Handling object permissions
+
+        def has_module_perms(self, user_obj, app_label):
+            return any(
+                perm[:perm.index('.')] == app_label
```
I would personally also prefer that, but this is the actual implementation from `ModelBackend`.

(I was thinking about moving this to BaseBackend, but there is no guarantee that third party backends use the same naming schema)

xi commented on docs/topics/auth/customizing.txt:
```diff
@@ -255,13 +255,18 @@ authenticated users do not.
+method and can return the object level permission as appropriate. When using
+:class:`~django.contrib.auth.mixins.PermissionRequiredMixin` you can implement
+:meth:`~django.contrib.auth.mixins.PermissionRequiredMixin.get_permission_object`
+to check for object permissions.
```
I understood this section as explaining object-level permissions in general, not only the backend part. I think it is crucial to explain how this can actually be used.

carltongibson replied:
OK: so if we show how to configure the backend, and add the new custom backend, then how about adding this towards the end, showing how it gets used? (Rather than here at the beginning — my thought being that until the custom backend is in place it's not functional/relevant.)

xi replied:
I guess you are right: It is probably sufficient to explain how this works with `user.has_perm()` here. I removed the part.

xi commented on docs/topics/auth/customizing.txt:
```diff
@@ -259,12 +259,48 @@ Handling object permissions
+The default :class:`~django.contrib.auth.backends.ModelBackend` only handles
+global permissions. That means that checking for object permissions will always
+return ``False`` or an empty set (depending on the check performed).
+Conversely, it is expected that any backend that handles object permissions
```
My idea here was to extend the specific behavior of `ModelBackend` to a genaral rule: Backends should handle either global or object permissions. For authors of global permission backends this means that they should always check `obj is None` (on object permission backends the equivalent `obj is something` needs to be checked anyway, so that is less of an issue).

Any idea where to put this instead? (do you even agree with that rule?)

carltongibson replied:
OK. Cor. I'm not sure (at all). 

I think I can easily imagine someone creating a single backend that uses object permissions but then falls back to the general permissions... — I'm not sure I'd want to legislate against that. 

(The permissions system has been ≈unchanged for ≈10 years: IMO one of it's strengths is not having strong opinions — I know people ask for the "superset" behaviour, so OK we can facilitate that but my bias would be towards less, rather than more...)

xi replied:
I understand your point about keeping it flexible. I removed this part and restored the original text.

xi commented:
Thanks for the extensive feedback!

I removed the last commit and incorporated some of your suggestions. As there were quite a lot, I may have forgotten something. Maybe you can mark some comments as resolved or add further feedback and I will do a second round of changes.

I am not completely sure how the review process usually works: Am I expected to amend/rebase my commits or add additional commits on top? Should I mark comments as resolved, should I comment with "done", or should I leave that completely to you? I am appy to adapt to any process you prefer :smiley:

carltongibson commented:
Hi @xi. Process isn't that important. We're just trying to make it as good as we can. 🙂 I'll take another look later.

carltongibson commented on docs/topics/auth/customizing.txt:
```diff
@@ -259,12 +259,48 @@ Handling object permissions
+
+    class BlogPostBackend:
+        def authenticate(self, request, **kwargs):
+            return None
```
This could be part of building the example — we have `ModelBackend` in play, which handles authentication. 

But (I'm getting there slowly... 🙂) we should be showing the `AUTHENTICATION_BACKENDS` setting (at each stage) as part of showing _how to use the system_ not **just** showing the backend implementation in a vacuum. (Especially given the new backend and permission mixin stuff — we need to show how they combine.)

carltongibson commented:
Oh, Once the docs are right...

I'm inclined to see this as a **single fix/improvement**: for example, I understand your point but, I can't really see the benefit of the `get_permissions_object()` hook if we don't add the rest. (Better leaving third-parties the API-surface-area.) 

As such, I'd check with Tim to see what he thinks but, I'd be inclined to pull it in as a single commit. At that point I might think the `BaseBackend` commit is worth adding as part of that too — if we go that way, it's part of the same improvement. (Not sure about that: just thinking.) 

Thanks again for your effort here! [#20218](https://code.djangoproject.com/ticket/20218) is 6 years old, so it's a nice win. 🙂

auvipy commented:
@paulayo https://docs.djangoproject.com/en/dev/internals/contributing/ follow this

xi commented:
Ok, I think I am now much closer to what you are thinking. I restored much of the documentation to its original state, only adding the incremental examples at the end.

Before we forget about it: I am not completely happy with the name "DefaultObjectPermissionBackend". Now is the last time we can change it. Any ideas? What about "InheritGlobalPermissionBackend"?

carltongibson commented:
Hi @xi. Sorry for the slow follow-up here, I needed a moment to get to this properly. 

I have comments relating to four points: 

* The docs changes here. 
* [#20218](https://code.djangoproject.com/ticket/20218) Adding DefaultObjectPermissionsBackend.
* [#29918](https://code.djangoproject.com/ticket/29918) PermissionRequiredMixin changes.
* The simplifications to the API you [originally discussed on the mailing list](https://groups.google.com/d/topic/django-developers/CNmz22gEsCc/discussion). 
    

**Docs Changes**

I (still) don't think these are clear enough. The `BlogPostBackend` seems misleading. It's seems contrived to justify the need for `DefaultObjectPermissionsBackend`, which only really arrises because of inter-op with django-guardian. 

For me you'd write it something like this: 
    
    
    def get_all_permissions(self, user_obj, obj=None):
        """Owner can edit/delete posts, others can only view"""
        perms = {'blog.view_blogpost'}
        if user_obj.is_active and not user_obj.is_anonymous:
            if isinstance(obj, BlogPost) and obj.owner == user_obj:
                perms += {'blog.change_blogpost', 'blog.delete_blogpost'}
        return perms

... and then the permission checks work as expected. As such, I don't think the additions are all that helpful. 


**Adding DefaultObjectPermissionsBackend**

Adding a third backend (ModelBackend, django-guardian's and then the new DefaultObjectPermissionsBackend) seems like overkill. Why aren't we just subclassing ModelBackend and using that **instead** of, rather than as well as, `ModelBackend`?

Something like...


    from django.contrib.auth.backends import ModelBackend


    class ImpliedObjectPermissionsBackend(ModelBackend):
        """
        Ignores `obj` parameter to permission checks to apply general permissions
        at object-level.
    
        Use in place of `ModelBackend` in settings.AUTHENTICATION_BACKENDS to avoid
        the need to check permissions twice when using Guardian:
    
            user.has_perm('foo.change_bar', obj=bar)
        
        Rather than: 
        
            user.has_perm('foo.change_bar', obj=bar) or user.has_perm('foo.change_bar')
        
        ...when using ModelBackend.
        """
        def has_perm(self, user, perm, obj=None):
            return super().has_perm(perm)
        
        ...
    
This would be a small little class, and since the need for this only comes up when using (e.g.) django-guardian I would suggest it would make an excellent contribution to `guardian.backends` there. 

I think if this were part of django-guardian the weirdness of explaining the use-case would also disappear. 

It seems to me it's natural home is there. I'll comment on the ticket as such. 


**PermissionRequiredMixin changes**

On review I realise that the code suggested here already exists in `guardian.mixins.PermissionRequiredMixin`. (I should have noticed this before.)

Rather than pulling this out of its natural home in django-guardian, we should just say "Use that". 

I'll comment on the ticket but, unless we're going to adopt a much more fully-featured implementation (which we're likely not) I think it's an error to try and pull bits and pieces in. 

In general, we probably want less code here rather than more, to wit...


**Simplifications to the API**

Your original ideas appear to have got lost somewere in the existing old tickets. 

Make `get_user_permissions()` public? (i.e. document it) — Yeah, why not? What other small changes? 

I think there's too much going on here to pursue those here. I think we should close this one and try afresh with new Trac tickets and/or PRs. I hope that seems fair. 

Closing on that basis but happy to discuss. Thanks!

xi commented:
Thanks again for the extensive feedback.

To be honest, I don't really get your argument. I am not sure why you think that django guardian is the only relevant backend. Agreed, many projects are not actively maintained (e.g. [django-permission](https://github.com/lambdalisue/django-permission)), but at least [django-rules](https://github.com/dfunckt/django-rules) is.

More importantly, django-guardian only covers a very specific usecase (checking permissions on `View.object`). So people might have to write their own backends. For example, I wrote a [custom backend for membership roles](https://git.mpib-berlin.mpg.de/castellum/castellum/blob/master/castellum/studies/backends.py) for a project I am working on. In order to do that I had to duplicate quite some code from other projects (such as django-guardian). In fact, every new backend will have to duplicate the exact same code.

Please don't get me wrong. I understand that this is a rarely used feature and you would prefer to keep it simple. I am just not sure how to continue: Do you think that django-guardian should be the only supported backend? Or do you think that custom backends should be supported? In that case I cannot follow your argument.

I could get into more detail about your comments, but I think this fundamental question needs to be answered first.

carltongibson commented:
Hi @xi. No problem with supporting custom backends. (The extension point is already there.) But django-guardian is by far the biggest beast in this area: I'm quite happy to recommend people in install it even if all they want to use are the permissions mixin and (if it were to be added) the new suggested `ModelBackend` subclass. 

Ultimately, having gone over this with you here extensively, I just can't see how these changes fit comfortably in Django itself, whereas adding them to Guardian (or in the case of the mixin, just leaving it where it is) is entirely natural. If it were code, we'd likely call the difficulty in spelling out the docs changes a _smell_ — i.e. a warning we weren't putting it in the right place. I think the same thing occurs here. 

For me, the bottom-line is that we've explicitly ruled this stuff out-of-scope and, so, given that sneaking in small parts of it just doesn't fit.

I hope that makes sense.

xi commented:
Even though I don't follow, I can accept your decision. Just let me try one final argument., maybe it convices you to think about it again:

What if I use django-guardian *and* django-rules in the same project? They both ship their own PermissionRequiredMixin, so which one am I supposed to use. In practice, either one will probably do the job, but to me this is a strong pointer that the mixin (and some other parts) actuelly *does* belong in django core.

In the case this does not convice you, do you want me to make a new pull request for any changes I proposed here or on the mailinglist? (You mentioned adding `get_user_permissions()` but as `get_group_permissions()` is already not being used I think that would be a step in the wrong direction)

carltongibson commented:
Hi @xi. I'm not anti-improvements here. 🙂 

I just don't think the suggestions in this PR, as we've arrived at, make sense. (They are, add the mixin from DG, the backend, and the docs example, which I think clouds the situation...) 

Now **maybe** you are right, and we should extend the support for object permissions in core. Maybe. 

But if so it should be feasible to come up with changes, starting with the docs I think, that are _"Yes, that's obviously much clearer"_. In such a situation I'm all for it. But I don't think we're there. 

For me, your initial email was "Let's improve/simplify the permissions API". So what does the new API look like, and what's the migration path to it? (This doesn't sound like a little project, but it could be a worthy one.)

At it stands, what I'm not keen on, is **just** pulling features out of a maintained third-party package to include them here. Unless there's an improvement on the cards, leave them where they are. 

> ... already not being used...

I don't have a suggestion in hand. (I don't mind the existing system.) But `get_group_permissions()` is used by `get_all_permissions()` which is used by `has_perm()`, which is the main entry point, so I'm not sure what you mean. 

Elsewhere, not here, blue skies: what do you think the permission system should look like? How is that clearer than what we have?

xi commented:
> But get_group_permissions() is used by get_all_permissions() which is used by has_perm(), which is the main entry point, so I'm not sure what you mean.

Yes, it is used internally. But it is not used as an extension point by either django-guardian or django-rules. (Note that django-guardian has a similar function somewhere internally, but it is not exposed on the backend)

I will try to create a new, clean merge request.
